### PR TITLE
Signature_lib: add of_string_exn/to_string + add a test for a special case for signature

### PR DESCRIPTION
**Pull request by dannywillems**
Signature_lib: Add to_string and of_string_exn methods with tests.
Add interface file for private_key.ml.
Add corner case for `String_sign.sign` to verify if the behavior is the same as https://github.com/o1-labs/proof-systems/pull/3305/commits/1bf7dfc948cd872924eed615e5f04a8c52e62554 (it is not, but the Caml implementation is the reference).

### CODE REVIEW

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-bypass-changelog

File: src/lib/string_sign/tests/test_string_sign.ml [hunk 1]
```diff
@@ -142,6 +142,26 @@ let test_other_network_failures () =
...
 
+let test_secret_key_between_scalar_field_and_base_field () =
+  (* There are 86663725065984043395317760 values between the two moduli.
+     - Base:
+       28948022309329048855892746252171976963363056481941560715954676764349967630337
+     - Scalar:
+       28948022309329048855892746252171976963363056481941647379679742748393362948097
+  *)
+  (* scalar modulus minus one *)
+  let sk_str =
```
Comment (martyall) [inline src/lib/string_sign/tests/test_string_sign.ml]: Why not generate a random value in the interval? Maybe in addition to `Base` and `Scalar - 1`?

File: src/lib/signature_lib/private_key.mli [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+(** Convert a private key to its decimal string representation.
+
+    Returns the private key value as a decimal string. This function is the
+    inverse of [of_string_exn], meaning that [of_string_exn (to_string t) = t] for any
+    valid private key [t].
+
+    @param t A private key to convert
+    @return A decimal string representation of the private key value
+
+    Examples:
```
Comment (martyall) [inline src/lib/signature_lib/private_key.mli]: I don't think these examples are needed, the comment about "inverses" above already states this.

File: src/lib/signature_lib/private_key.mli [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    The string must represent a non-negative integer in decimal format that is
+    strictly less than the field modulus
+    (28948022309329048855892746252171976963363056481941647379679742748393362948097).
+
+    @param s A decimal string representation of a private key value
+    @return A private key corresponding to the given value
+    @raise Failure if the string is not a valid decimal integer or if the value
+    is greater than or equal to the field modulus
+
+    Examples:
```
Comment (martyall) [inline src/lib/signature_lib/private_key.mli]: I don't think these examples are needed

File: src/lib/signature_lib/test/test_private_key.ml [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        let modulus = Bignum_bigint.of_string modulus_string in
+        let value = Bignum_bigint.of_string s in
+        if Bignum_bigint.(value >= modulus) then
+          Alcotest.(check bool)
+            ("Value " ^ s ^ " >= modulus raises exception")
+            true true
+        else Alcotest.fail ("Unexpected exception for valid value " ^ s) )
+
+let test_of_string_exn_edge_case_empty () =
+  (* Test edge case: empty string should be treated as zero *)
```
Comment (martyall) [inline src/lib/signature_lib/test/test_private_key.ml]: This comment feels wrong, the test is stating that the empty string throws an exception?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+       Snark_params.Tick.Inner_curve.Scalar.zero )
+
+let test_of_string_exn_one () =
+  let key = Private_key.of_string_exn "1" in
+  Alcotest.(check bool)
+    "Private key from string \"1\" should create valid key" true
+    (Snark_params.Tick.Inner_curve.Scalar.equal key
+       Snark_params.Tick.Inner_curve.Scalar.one )
+
+let test_of_string_exn_small_values () =
```
Comment (martyall) [inline src/lib/signature_lib/test/test_private_key.ml:33]: generate random `Int` values and use the builtin `int_to_string` ?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+  try
+    let _key = Private_key.of_string_exn two_times_modulus_plus_123 in
+    Alcotest.fail "Values >= modulus should raise an exception"
+  with _ ->
+    Alcotest.(check bool) "Values >= modulus raise exception" true true
+
+let test_of_string_exn_random_values () =
+  (* Test with some random large values less than modulus *)
+  let random_values =
+    [ "12345678901234567890123456789012345678901234567890123456789012345678901234567"
```
Comment (martyall) [inline src/lib/signature_lib/test/test_private_key.ml]: use the `modulus` constant and generate random offsets?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+let test_of_string_exn_edge_case_invalid () =
+  (* Test edge case: invalid string *)
+  try
+    let _key = Private_key.of_string_exn "not_a_number" in
+    Alcotest.fail "Invalid string should raise an exception"
+  with _ -> Alcotest.(check bool) "Invalid string raises exception" true true
+
+let test_of_string_exn_negative_values () =
+  (* Test negative values - should raise exceptions *)
+  let negative_values = [ "-1"; "-42"; "-100"; "-1000000000" ] in
```
Comment (martyall) [inline src/lib/signature_lib/test/test_private_key.ml]: generate random negative `Int` values and use builtin `int_to_string`?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+        let _key = Private_key.of_string_exn s in
+        Alcotest.fail ("Invalid format " ^ s ^ " should raise an exception")
+      with _ ->
+        Alcotest.(check bool)
+          ("Invalid format " ^ s ^ " raises exception")
+          true true )
+
+let test_to_string_basic () =
+  (* Test basic to_string functionality *)
+  let test_values = [ "0"; "1"; "2"; "10"; "42"; "100"; "1000" ] in
```
Comment (martyall) [inline src/lib/signature_lib/test/test_private_key.ml]: generate random `Int` values?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+      let key = Private_key.of_string_exn s in
+      let result = Private_key.to_string key in
+      Alcotest.(check string)
+        ("to_string should return correct value for " ^ s)
+        s result )
+
+let test_to_string_large_values () =
+  (* Test to_string with large values *)
+  let large_values =
+    [ "12345678901234567890123456789012345678901234567890123456789012345678901234567"
```
Comment (martyall) [inline src/lib/signature_lib/test/test_private_key.ml]: similar comment as the others to generate random values

File: src/lib/signature_lib/test/test_private_key.ml [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+        if Bignum_bigint.(value >= modulus) then
+          Alcotest.(check bool)
+            ("Large value " ^ s ^ " >= modulus as expected")
+            true true
+        else Alcotest.fail ("Unexpected exception for valid large value " ^ s) )
+
+let test_roundtrip_of_string_exn_to_string () =
+  (* Test that of_string_exn and to_string are inverse functions *)
+  let test_values =
+    [ "0"
```
Comment (martyall) [inline src/lib/signature_lib/test/test_private_key.ml]: generate random values

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
@martyall I think I implemented all your requested changes. Can you check again please?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+  let large_values = modulus_minus_one :: random_large_values in
+  List.iter large_values ~f:(fun s ->
+      try
+        let key = Private_key.of_string_exn s in
+        let result = Private_key.to_string key in
+        Alcotest.(check string)
+          ("to_string should return correct value for large number " ^ s)
+          s result
+      with _ ->
+        (* If the value is >= modulus, of_string_exn should raise an exception *)
```
Comment (glyh) [inline src/lib/signature_lib/test/test_private_key.ml]: Why will this branch be reached?

Comment (glyh) [reply src/lib/signature_lib/test/test_private_key.ml]: modules minus offset should always be smaller than modulus, no?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+  List.iter test_values ~f:(fun original ->
+      try
+        let key = Private_key.of_string_exn original in
+        let roundtrip = Private_key.to_string key in
+        Alcotest.(check string)
+          ( "Roundtrip of_string_exn -> to_string should preserve value for "
+          ^ original )
+          original roundtrip
+      with _ ->
+        (* If the value is >= modulus, of_string_exn should raise an exception *)
```
Comment (glyh) [inline src/lib/signature_lib/test/test_private_key.ml]: Why this branch will be reached?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+        let value = Bignum_bigint.of_string original in
+        if Bignum_bigint.(value >= modulus) then
+          Alcotest.(check bool)
+            ("Value " ^ original ^ " >= modulus as expected")
+            true true
+        else Alcotest.fail ("Unexpected exception for valid value " ^ original) )
+
+let test_roundtrip_to_string_of_string_exn () =
+  (* Test that to_string -> of_string_exn is also inverse for generated keys *)
+  let gen_key1 = Private_key.of_string_exn "42" in
```
Comment (glyh) [inline src/lib/signature_lib/test/test_private_key.ml]: Nit: I think you can use generators to generate strings of integers.

File: src/lib/signature_lib/test/test_private_key.ml [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+        (* If the value is >= modulus, of_string_exn should raise an exception *)
+        let modulus = Bignum_bigint.of_string modulus_string in
+        let value = Bignum_bigint.of_string s in
+        if Bignum_bigint.(value >= modulus) then
+          Alcotest.(check bool)
+            ("Large value " ^ s ^ " >= modulus as expected")
+            true true
+        else Alcotest.fail ("Unexpected exception for valid large value " ^ s) )
+
+let test_roundtrip_of_string_exn_to_string () =
```
Comment (glyh) [inline src/lib/signature_lib/test/test_private_key.ml:210]: This is rather confusing. Previous test is already roundtripping, I think?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+    (Snark_params.Tick.Inner_curve.Scalar.equal key
+       Snark_params.Tick.Inner_curve.Scalar.one )
+
+let test_of_string_exn_small_values () =
+  let gen = Generator.int_inclusive 1 1000000 in
+  let test_values =
+    List.init 5 ~f:(fun _ -> Quickcheck.random_value gen |> Int.to_string)
+  in
+  List.iter test_values ~f:(fun s ->
+      let key = Private_key.of_string_exn s in
```
Comment (glyh) [inline src/lib/signature_lib/test/test_private_key.ml:37]: So these 2 functions are just the same function? If so why even bother design them to be 2 functions. instead of alias of each other?

File: src/lib/signature_lib/private_key.ml [hunk 16]
```diff
@@ -66,3 +64,7 @@ let of_yojson = function
           Error ("Signature_lib.Private_key.of_yojson: " ^ Exn.to_string exn) )
   | _ ->
       Error "Signature_lib.Private_key.of_yojson: Expected a string"
+
+let of_string_exn s = Inner_curve.Scalar.of_string s
+
+let to_string t = Inner_curve.Scalar.to_string t
```
Comment (glyh) [inline src/lib/signature_lib/private_key.ml:70]: I personally don't see the value of aliasing a function. This only creates confusion. downstream user would always need to check the underlying implementation to ensure they're the same thing.

Unless this is used as part of functor calculation, I'd like to reject such change.

File: src/lib/signature_lib/test/test_private_key.ml [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+  List.iteri test_keys ~f:(fun i key ->
+      let str_repr = Private_key.to_string key in
+      let roundtrip_key = Private_key.of_string_exn str_repr in
+      Alcotest.(check bool)
+        ( "Roundtrip to_string -> of_string_exn should preserve key "
+        ^ Int.to_string i )
+        true
+        (Snark_params.Tick.Inner_curve.Scalar.equal key roundtrip_key) )
+
+let test_to_string_modulus_minus_one () =
```
Comment (glyh) [inline src/lib/signature_lib/test/test_private_key.ml]: This is covered in `test_to_string_large_values` already. Is this AI generated code?

File: src/lib/signature_lib/test/test_private_key.ml [hunk 18]
```diff
@@ -0,0 +1,312 @@
+open Core_kernel
+open Signature_lib
+open Base_quickcheck
+
+let modulus_string =
+  "28948022309329048855892746252171976963363056481941647379679742748393362948097"
```
Comment (glyh) [inline src/lib/signature_lib/test/test_private_key.ml]: I believe these constants are defined somewhere else. If possible please avoid copying the same number around. This would make our lives easier.

File: src/lib/string_sign/tests/test_string_sign.ml [hunk 19]
```diff
@@ -142,6 +143,36 @@ let test_other_network_failures () =
...
+let test_secret_key_between_scalar_field_and_base_field () =
+  (* There are 86663725065984043395317760 values between the two moduli.
+     - Base:
+       28948022309329048855892746252171976963363056481941560715954676764349967630337
+     - Scalar:
+       28948022309329048855892746252171976963363056481941647379679742748393362948097
+  *)
+  let base_modulus =
+    Bignum_bigint.of_string
+      "28948022309329048855892746252171976963363056481941560715954676764349967630337"
```
Comment (glyh) [inline src/lib/string_sign/tests/test_string_sign.ml]: Is this randomly chosen, or defined somewhere else? 
If first case, I'd use PBT; If second, I'd make this a reference to that definition.

File: src/lib/string_sign/tests/test_string_sign.ml [hunk 20]
```diff
@@ -142,6 +143,36 @@ let test_other_network_failures () =
...
+    |> Int64.to_string |> Bignum_bigint.of_string
+  in
+  let sk_bignum = Bignum_bigint.(base_modulus + random_offset) in
+  let sk_str = Bignum_bigint.to_string sk_bignum in
+
+  let secret_key = Signature_lib.Private_key.of_string_exn sk_str in
+  let keypair = Signature_lib.Keypair.of_private_key_exn secret_key in
+
+  let s = "Rain and Spain don't rhyme with cheese" in
+  let signature = sign ~signature_kind:Mainnet keypair.private_key s in
```
Comment (glyh) [inline src/lib/string_sign/tests/test_string_sign.ml]: Usually we use testnet sig kind for all tests. 
If we just don't care sig kind, I'd suggest we generate a sig kind randomly.

**Review by glyh — CHANGES_REQUESTED** [review]

File: src/lib/signature_lib/test/test_private_key.ml (hunk 11 above)
Comment (dannywillems) [reply src/lib/signature_lib/test/test_private_key.ml]: Removed.

File: src/lib/signature_lib/test/test_private_key.ml (hunk 12 above)
Comment (dannywillems) [reply src/lib/signature_lib/test/test_private_key.ml]: Removed.

File: src/lib/signature_lib/test/test_private_key.ml (hunk 13 above)
Comment (dannywillems) [reply src/lib/signature_lib/test/test_private_key.ml]: Done.

File: src/lib/signature_lib/test/test_private_key.ml (hunk 14 above)
Comment (dannywillems) [reply src/lib/signature_lib/test/test_private_key.ml:210]: But only one way. Here we test the other inverse. We want to test that `f o g = g o f = id`. Therefore we need two tests.

File: src/lib/signature_lib/private_key.ml (hunk 16 above)
Comment (dannywillems) [reply src/lib/signature_lib/private_key.ml:70]: The type is not the same. The user would need to know that `Inner_curve.Scalar.t` is equal to `Private_key.t`, which is not an assumption we should make when we design an API.
A user of the library would probably check first the interface file to know which method we allow on the type itself. The user should not use any other routines than the module authorizes/exposes.

File: src/lib/signature_lib/test/test_private_key.ml (hunk 17 above)
Comment (dannywillems) [reply src/lib/signature_lib/test/test_private_key.ml]: Removed. It was duplicated later. There was only this method before.

> Is this AI generated code?

It has been partially generated with Claude for boilerplate, yes.

File: src/lib/string_sign/tests/test_string_sign.ml (hunk 19 above)
Comment (dannywillems) [reply src/lib/string_sign/tests/test_string_sign.ml]: > Is this randomly chosen, or defined somewhere else?

No. As reflected in the variable name, it is the modulus of the base field of Pallas, the curve used for signing.

File: src/lib/string_sign/tests/test_string_sign.ml (hunk 20 above)
Comment (dannywillems) [reply src/lib/string_sign/tests/test_string_sign.ml]: Updated.

File: src/lib/string_sign/tests/test_string_sign.ml (hunk 19 above)
Comment (dannywillems) [reply src/lib/string_sign/tests/test_string_sign.ml]: See 31193786a7bac645fd05e44c5ed266f2903f991f

File: src/lib/signature_lib/test/test_private_key.ml (hunk 18 above)
Comment (dannywillems) [reply src/lib/signature_lib/test/test_private_key.ml]: See 010d891a78d799401099198430756211fffdbf1c

File: src/lib/signature_lib/test/test_private_key.ml (hunk 15 above)
Comment (dannywillems) [reply src/lib/signature_lib/test/test_private_key.ml:37]: > So these 2 functions are just the same function?

Wdym?

**Comment by dannywillems** [comment]
!ci-build-me

File: src/lib/signature_lib/test/test_private_key.ml (hunk 15 above)
Comment (glyh) [reply src/lib/signature_lib/test/test_private_key.ml:37]: Looks like this comment is in wrong place. I'm talking about the fact we're aliasing the functions and exposing them. But it seems you explained it

**Comment by dannywillems** [comment]
Bumping approvals from @martyall and @glyh and merge.
